### ISSUE TITLE
Rollback aware auto-sync

### ISSUE THREAD

**Issue by stereobutter** [argoproj/argo-cd#9570]
# Summary

Currently auto-sync and rollback do not play nicely together. To perform a rollback auto-sync has to be disabled first and if reenabled happily will sync again to the commit that was rolled back previously. If argo remembered the commit that got rolled back and excluded it from auto-sync this would improve that situation a lot. 

This might be the same or a similar issue as https://github.com/argoproj/argo-cd/issues/5351 but I found the issue description really unclear. 

# Motivation
What has to happen when a rollback is needed currently entails multiple steps to get an application back on track:

1. disable auto-sync
2. perform rollback
3. revert/fix the offending commit in git
4. reenable auto-sync

If auto-sync excluded the offending commit this would be reduced to
1. perform rollback
2. revert/fix the offending commit in git

Apart from saving a developer a bit of manual effort, where this feature would be really useful is for automating rollbacks. We currently tinker with using a post-sync hook for running some tests against our application and would like to automatically rollback using a fail hook (via the argo api [^1]). When a developer then releases a fix, the application would automatically sync again and everything would be back to normal.

[^1]: it appears this would currently not work without some level of indirection due to https://github.com/argoproj/argo-cd/issues/6494. Maybe one could use a argo-workflow that starts another workflow to do the rollback?

# Proposal
Argo would grow a new option that when enabled would cause auto-sync to ignore a commit if it has been rolled back. For the basic use case it would probably suffice to just remember the last commit that was rolled back (compared to keeping a history of commits).

**Comment by alexef** [comment]
linking https://github.com/argoproj/argo-cd/issues/8686 here as a way to connect the dots of all these requests

**Comment by jsirianni** [comment]
This behavior surprised me. Right now, rolling back feels unsafe because Argo will happily re-deploy shortly after.

**Comment by afernandez-01** [comment]
+1

**Comment by selcuktemizsoy** [comment]
Is there any update on this topic? We have some argo workflows where we need this improvement. If someone solved it already, I will be really appreciated. 

Our workflow: 
Run argo cd sync
Run some tests after argo cd sync as post sync hook
After this step we want to have auto rollback where argo cd do this automatically if post sync fails. 

simply: Deploy-Post-sync if fails rollback if pass do nothing.

**Comment by Kampe** [comment]
Would also like to know why these two tools do not play well together on rollback..... when you need them to.

**Comment by chinkung** [comment]
Rollback button should be disabled if auto-sync is enabled

**Comment by AndresPinerosZen** [comment]
@chinkung With the current behavior, yes. Having the rollback button enabled can confusing. But if Git reconciliation is implemented for the rollback, then it should be enabled all the time. They are asking to implement the reconciliation in Git automatically, which makes sense tbh. Just because it is GitOps doesn't mean that there can't be nice UI features to change the source of truth.

**Comment by jluque0101** [comment]
This issue affects us too. Users have to go through multiple steps, including creating PRs to disable auto-sync, just to perform a successful rollback without overwriting changes.

**Comment by StyleTang** [comment]
Hi Team,

Can we provide a mechanism where certain parameters are not controlled by Yaml auto-sync and are only controlled by the UI? 

For example, we can add an UI control only parameter `syncPolicy.uiControl.blockAutomatedSync` (default false).

Assuming that our service is enabled with syncPolicy, the final configuration would be:

```yaml
syncPolicy:
    uiControl:
      blockAutomatedSync: false
    automated:
      prune: true
      selfHeal: true
```

The user story would be as follows:

* 1. When the user clicks on rollback in the UI, `blockAutomatedSync` is automatically set to `true`.
* 2. Rollback to the specified version.
* 3. Periodic/manual sync is triggered, and the syncPolicy mechanism detects the difference, but since `blockAutomatedSync` is `true`, it blocks sync.
* 4. After the bug is fixed and the latest code is committed, it will still not sync.
* 5. The user clicks "unblock automatedSync" in the UI and sets `blockAutomatedSync` to `false`.
* 6. Periodic/manual sync is triggered and the latest version is deployed.

Is this feasible, or does anyone have concerns or better suggestions?

Thanks！

**Comment by billyatroadie** [comment]
Here for the notifications

**Comment by andrii-korotkov-verkada** [comment]
Right now using rollback from History and Rollback would prompt you to disable autosync, reducing a number of steps by one and mitigating concerns about re-sync. Please, let me know if this proposal is still needed.

**Comment by yohanb** [comment]
> Right now using rollback from History and Rollback would prompt you to disable autosync, reducing a number of steps by one and mitigating concerns about re-sync. Please, let me know if this proposal is still needed.

Hi! Are you saying that this is a current feature in 2.13?

Another problem with this approach is applications that are created through application sets. Even if you disable auto-sync, it's re-applied by the appset.

**Comment by andrii-korotkov-verkada** [comment]
> Hi! Are you saying that this is a current feature in 2.13?

It's been around for some time for applications. I don't know about application sets though.

**Comment by shameemshah** [comment]
Here for the notifications

**Comment by MaenIb** [comment]
Hello, any update on this issue ?

My case involves the use of ApplicationSets for all of our application deployments, so even from the UI I can't disable the auto sync (it re-enable the auto-sync automatically), I assume this happens because the appset sees that there is a state change so it tries to redeploy the state that is defined in the appset definition (if I'm not mistaken).

Is there any work-around to deal with this, instead of using the CLI each time to disable the auto-sync in the appset and then having to reenable it again, after the assumed bug (or whatever) is fixed ?

Thanks!

**Comment by rim-b-nasr** [comment]
Here for the notifications

**Comment by ddragoti** [comment]
@MaenIb the only way I've found is adding the syncPolicy to the ignoreDifferences block like this:

```
spec:
  ignoreApplicationDifferences:
    - jsonPointers:
        - /spec/syncPolicy
```

But it then requires having to remember to turn Autosync on manually afterwards.

**Comment by kostis-codefresh** [comment]
> If auto-sync excluded the offending commit this would be reduced to
> 
>     1. perform rollback
>     2. revert/fix the offending commit in git
> 
> Apart from saving a developer a bit of manual effort, where this feature would be really useful is for automating rollbacks. We currently tinker with using a post-sync hook for running some tests against our application and would like to automatically rollback using a fail hook 

This seems to me functionality already covered by [Argo Rollouts](https://argoproj.github.io/rollouts/) and its [Analysis object](https://argo-rollouts.readthedocs.io/en/stable/features/analysis/).

And Argo Rollouts works just fine with applications that use auto-sync.

If you use Argo Rollouts the list actually becomes just one item (for humans)

1. Argo Rollouts performs rollback on its own (no human intervention here)
2. A developer reverts/fixes the offending Git commit in Git

And this works with auto-sync/self-heal enabled (as it should be)

The Argo Rollouts documentation [contains more information](https://argo-rollouts.readthedocs.io/en/stable/FAQ/#what-is-the-relationship-between-rollbacks-with-argo-rollouts-and-rollbacks-with-argo-cd).

**Comment by Tiiberiu** [comment]
If auto-sync is on, can we have possibility to make a custom rollback action? I would like to do a api request to trigger my pipeline that rollbacks by commit

**Comment by nh4ttruong** [comment]
I think the best practice for handling auto-sync in ApplicationSets is to enable rollback processes at the application level, rather than relying on the risky approach of reverting a commit hash. This method simplifies and speeds up the rollback process while allowing temporary rollbacks without the need to write back to Git or an additional step in CI/CD.

> Assuming that our service is enabled with syncPolicy, the final configuration would be:
> 
> ```
> syncPolicy:
>     uiControl:
>       blockAutomatedSync: false
>     automated:
>       prune: true
>       selfHeal: true
> ```
> The user story would be as follows:
> 
> * 1. When the user clicks on rollback in the UI, `blockAutomatedSync` is automatically set to `true`.
> * 2. Rollback to the specified version.
> * 3. Periodic/manual sync is triggered, and the syncPolicy mechanism detects the difference, but since `blockAutomatedSync` is `true`, it blocks sync.
> * 4. After the bug is fixed and the latest code is committed, it will still not sync.
> * 5. The user clicks "unblock automatedSync" in the UI and sets `blockAutomatedSync` to `false`.
> * 6. Periodic/manual sync is triggered and the latest version is deployed.

**Comment by ppapapetrou76** [comment]
here's a summary of the issue and the topics discussed do far 


TL;DR: This issue requests that Argo CD be aware of rollbacks when auto-sync is enabled, so that manually rolling back an application is not immediately overridden by the next auto-sync cycle.

---
The Core Problem

Today, auto-sync and rollback are mutually exclusive by design: if auto-sync is enabled, triggering a rollback returns a hard error (rollback cannot be initiated when auto-sync is enabled). This forces the following workflow every time a rollback is needed:

1. Disable auto-sync
2. Perform the rollback
3. Revert/fix the offending commit in Git
4. Re-enable auto-sync

The original issue proposes that Argo CD should remember which revision was rolled back and have auto-sync skip that revision until a new, different revision is pushed to Git — reducing the workflow to just steps 2 and 3.

---
Key Topics from the Discussion

1. The ApplicationSet complication

Several users pointed out that for applications managed by an ApplicationSet, even the current workaround breaks down: disabling auto-sync on the Application is immediately overridden by the ApplicationSet controller restoring the original sync policy. This makes rollbacks in ApplicationSet-managed environments particularly painful. A workaround using `ignoreApplicationDifferences` for `/spec/syncPolicy` was suggested, but it requires manually re-enabling auto-sync afterward.

2. Partial UI improvement already shipped

The History & Rollback UI dialog now prompts users to disable auto-sync before proceeding with a rollback. This reduces the manual steps slightly but does not address the core request: rollback and auto-sync working together seamlessly.

3. Automating rollbacks is still broken

The real value of this feature is enabling automated rollback pipelines (e.g., via a post-sync hook or an Argo Workflow). Without rollback-aware auto-sync, any automated rollback will be undone immediately, making that pattern unreliable.

4. Community workaround: inspecting sync history programmatically

There is a Go-based workaround that compares `status.sync.revision` with `status.operationState.operation.sync.revision` against the deployment history to detect a rolled-back state and abort a sync pipeline. This is functional but requires significant custom tooling.

5. Proposed solution: a `blockAutomatedSync` flag

There is an idea to add a new field (e.g., `syncPolicy.uiControl.blockAutomatedSync`) that would be automatically set to true when a rollback is triggered and cleared manually by the user once a fix is in Git. This is simpler to implement but still requires a manual unblock step.

6. Alternative: use Argo Rollouts

Argo Rollouts already handles automated rollbacks and works correctly with auto-sync enabled. For teams already using Rollouts, this issue may not apply.

**Current Code Behavior**

- Application Server hard blocks rollback if auto-sync is enabled.
- Application Controller with auto-sync enabled never checks whether the target revision was previously rolled back. It will re-sync to whatever targetRevision the spec declares.
- RevisionHistory type — no field to mark a revision as "rolled back"; it only stores revision, source, timestamp, and InitiatedBy.
I've analyzed the codebase and want to share a concrete implementation plan before opening a PR.


---
Proposed design

Store the "bad" revision (the one the user is rolling back from) in a new `app.Status.RolledBackRevision` field. Auto-sync reads that field and skips if the current desired revision matches. The field is cleared automatically when a subsequent successful sync deploys any other revision.

Because this lives in the status subresource, ApplicationSet reconciliation — which only overwrites spec — will never touch it. This also fixes the ApplicationSet use case 

The rollback flow with this change becomes:

1. User clicks rollback (auto-sync remains enabled)
2. Argo CD records the bad commit SHA in `app.Status.RolledBackRevision`
3. Auto-sync fires, sees OutOfSync, checks the guard → skips (bad SHA matches)
4. Developer pushes a fix commit (new SHA)
5. Auto-sync fires, new SHA does not match → syncs normally
6. Successful sync clears `app.Status.RolledBackRevision`

The fail / edge cases
- Rollback operation fails ->  `RolledBackRevision` field is never set, auto-sync continues normally
- User manually syncs to the "bad" revision after rollback -> The `RolledBackFromRevision` is not set on a manual sync op
- ApplicationSet overwrites `spec.syncPolicy` -> no effect. Appset code is unrelated to `app.status`
- Multiple successive rollbacks -> Last rollback wins; field is overwritten each time
- User disables then re-enables auto-sync without pushing a fix -> `RolledBackRevision` is written with the "bad revision"  — auto-sync stays blocked until git moves forward or rollback is eventually completed.
